lightning/src/ln/peer_handler.rs

```diff
@@ -1593,17 +1609,18 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref, CMH: Deref> P
 	/// an indication that TCP sockets have stalled even if we weren't around to time them out
 	/// using regular ping/pongs.
 	pub fn disconnect_all_peers(&self) {
-		let mut peers_lock = self.peers.lock().unwrap();
+		let mut peers_lock = self.peers.write().unwrap();
+		let mut node_id_to_descriptor = self.node_id_to_descriptor.lock().unwrap();
 		let peers = &mut *peers_lock;
 		for (mut descriptor, peer) in peers.peers.drain() {
-			if let Some(node_id) = peer.their_node_id {
+			if let Some(node_id) = peer.lock().unwrap().their_node_id {
```

Comment:
Oops. So this isn't an actual issue because we have the `peers` *write* lock, but there is a real inversion in the message handler. I built up more lockorder inversion tests in #1420 which I'll rebase this PR on, as well as some new tests here to hit the code in question.

Reply:
Actually, I don't need to base it on #1420, the new tests catch the two issues I see either way, so I'd say lets move forward without 1420.